crates/codegen/schema/src/validation/rules/references.rs

```diff
@@ -40,59 +43,56 @@ impl Visitor for ReferencesCollector<'_> {
-            let Node { cst_node, value } = reference;
-            if self.definitions.internal_named_parsers.contains_key(value) {
-                reporter.report(cst_node, Errors::NodesCannotBeReferenced(value.to_owned()))
-            } else if !self.definitions.top_level_parsers.contains_key(value)
-                && !self.definitions.top_level_scanners.contains_key(value)
-            {
-                reporter.report(cst_node, Errors::NoSuchProduction(value.to_owned()))
-            } else {
-                self.references.insert(reference.value.to_owned());
+        match &parser.definition.value {
+            ParserDefinition::Reference(reference) => {
+                let Node { cst_node, value } = reference;
+                self.check_is_scanner_or_parser_reference(cst_node, value, reporter);
+                return VisitorResponse::StepOut;
             }
-
-            return VisitorResponse::StepOut;
-        } else {
-            return VisitorResponse::StepIn;
+            ParserDefinition::DelimitedBy { open, close, .. } => {
+                let Node { cst_node, value } = &open.value.reference;
+                self.check_is_scanner_reference(cst_node, value, reporter);
+                let Node { cst_node, value } = &close.value.reference;
+                self.check_is_scanner_reference(cst_node, value, reporter);
+            }
+            ParserDefinition::SeparatedBy {
+                separator: reference,
+                ..
+            }
+            | ParserDefinition::TerminatedBy {
+                terminator: reference,
+                ..
+            } => {
+                let Node { cst_node, value } = &reference.value.reference;
+                self.check_is_scanner_reference(cst_node, value, reporter);
+            }
+            _ => {}
         }
+        return VisitorResponse::StepIn;
     }
 
-    fn visit_primary_expression_reference(
+    fn visit_precedence_parser(
         &mut self,
-        reference: &Reference,
+        parser: &PrecedenceParserRef,
         reporter: &mut Reporter,
     ) -> VisitorResponse {
-        let Node { cst_node, value } = &reference.reference;
-        if self.definitions.internal_named_parsers.contains_key(value) {
-            reporter.report(cst_node, Errors::NodesCannotBeReferenced(value.to_owned()))
-        } else if !self.definitions.top_level_parsers.contains_key(value) {
-            reporter.report(cst_node, Errors::NoSuchPrimaryExpression(value.to_owned()))
-        } else {
-            self.references.insert(value.to_owned());
-        }
-
-        return VisitorResponse::StepOut;
+        parser
+            .definition
+            .primary_expressions
+            .iter()
+            .for_each(|expression| {
+                let Node { cst_node, value } = &expression.value.reference;
+                self.check_is_scanner_or_parser_reference(cst_node, value, reporter);
+            });
+        return VisitorResponse::StepIn;
     }
 }
```

Comment:
I think `check_unused` has another bug, where it is not checking `self.definitions.top_level_scanners` as well.

Reply:
fixed